MANUAL.md

```diff
@@ -0,0 +1,194 @@
+    // Check the expected result
+    let snapshot = testkit.snapshot();
+    let schema = MySchema::new(&snapshot);
+    assert!(schema.is_my_data_checked());
+}
+```
+
+Make sure that you have full coverage of the business logic in the `execute` method of your transactions.
+But if you just want to check the `verify` logic in the transaction, you can do it without testkit in
+a simple way like this:
+
+```rust
+let tx = MyTransaction::new(...);
+assert!(tx.verify());
+```
+
+### Api testing
+
+The following steps may help you.
+
+* Define the `MyServiceApi` trait for the `TestKitApi` structure that covers the whole api of your service.
+* Implement functions that use transactions to fill your storage with the test data.
+* Create the tests that check all of your endpoints.
+
+```rust
+
+// Api trait definition.
+
+trait MyServiceApi {
+    fn get_public_data(&self) -> ApiResponsePublicData;
+    fn get_private_data(&self) -> ApiResponsePrivateData;
+    fn post_private_data(&self, data: MyPrivateData) -> ApiResponsePostPrivateData;
+}
+
+impl MyServiceApi for TestKitApi {
+    fn get_public_data(&self) -> ApiResponsePublicData {
+        self.get(ApiKind::Service("my_service"), "/v1/first_endpoint")
+    }
+
+    fn get_private_data(&self) -> ApiResponsePublicData {
+        self.get_private(ApiKind::Service("my_service"), "/v1/second_endpoint")
+    }
+
+    fn post_private_data(&self, data: &MyPrivateData) -> ApiResponsePostPrivateData {
+        self.post(
+            ApiKind::Service("my_service"),
+            "v1/third_endpoint",
+            &data,
+        )
+    }
+}
+
+// Api test skeleton
+
+#[test]
+fn my_api_test() {
+    let mut testkit = TestKitBuilder::validator()
+        .with_service(MyService::new())
+        .create();
+    fill_storage_with_data(&mut testkit);
+    // Check api responses
+    let api = testkit.api();
+    assert_eq!(api.get_public_data(), ApiResponsePublicData::new(...));
+    ...
+}
+```
+
+## Advanced usage
+
+Here are examples of more complex and less common cases.
+
+### Oracles testing
+
+The oracle in this case is service which can produce transaction with external data after the commit of block,
+[`exonum-time`][exonum-time] and [`exonum-btc-anchoring`][exonum-btc-anchoring] are examples of these kind of oracles.
+In this way, transactions which were created during the `handle_commit` execution will be stored in `TestKit` memory pool.
+
+```rust
+let mut testkit = TestKitBuilder::validator()
+    .with_service(HandleCommitService)
+    .create();
+// Check that `handle_commit` invoked on the correct height.
+for i in 1..5 {
+    testkit.create_block();
+    let tx = TxAfterCommit::new_with_signature(Height(i), &Signature::zero());
+    assert!(testkit.mempool().contains_key(&tx.hash()));
+}
+```
+
+In order to invoke a `handle_commit` event, you must create a block.
+If the oracle has to fetch any data from external world, you must create a mock object.
```

Comment:
This presupposes that service implementations must support mock objects, right? 😄  Maybe write it explicitly that if one designs a service interacting with the external world, he should allow to mock external APIs (possibly with recommendations how to do that and/or where to see the examples)?

Reply:
Yes, but this guide is not a collection of best practices. And I suppose that we have to write best practice collection later.